Implement Notification handling

lpsmith opened the pull request:
This pull request is dependent upon #42, of course.

lpsmith committed:
Implement notification handling

lpsmith committed:
Use threadWaitReadSTM when available

begriffs commented:
Looking forward to this being polished up and merged. It's important for https://github.com/begriffs/postgrest/issues/475

nikita-volkov commented:
I'm currently swamped. It'll require me at least another week to come back to this.

begriffs commented:
No problem, don't mean to stress you out. Wanted to post the update here so @lpsmith also knows I appreciate the PR and that it will be helpful once merged.

lpsmith committed:
Add comments explaining the reasoning behind the CPP cases

begriffs commented:
Do you have some time to review this PR?

nikita-volkov commented:
First I need to fix the bug with multithreading, which you know of, then I plan to switch to this issue. Can't make promises on the terms though. A month seems as likely as a week :)

nikita-volkov commented:
Okay, I've reviewed this. I must say I'm rather conflicted. Everything that gets done in this PR actually only needs "postgresql-libpq", and I've [actually refactored it to be so](https://github.com/nikita-volkov/hasql/blob/b45e71dfe45eb925b088004734fa3085fe687b21/library/Hasql/Notification.hs). This then raises the question of whether it makes sense at all to put this in the "hasql" library.

I think the best way to do this is for me to expose the low-level "postgresql-libpq" connection, and for the functionality of this PR to be released as a general extension-library over "postgresql-libpq", which itself won't have any dependency on Hasql. Thus it'll become possible to use this functionality with Hasql as well as any other libpq-based Postgresql driver.

What do you guys say?

lpsmith commented:
There is a dependency between `getNotification` and a higher-level binding:  your update has drastically changed the concurrent semantics for the worse.   You can no longer wait for notifications in one thread while also continuing to use that connection for (short) transactions in another thread.    There really isn't a generic way of implementing `getNotification` properly without dealing with issues that postgresql-libpq purposefully doesn't deal with.

nikita-volkov commented:
> Your update has drastically changed the concurrent semantics for the worse.

Right. I see now that the connection needs to be released in-between the loop cycles. However,

> There really isn't a generic way of implementing getNotification properly 

There is. The "postgresql-libpq"-based library could implement this using a more general interface like this:

```haskell
getNotification ::
  (forall a. (PQ.Connection -> IO a) -> IO a) ->
  IO (Either IOError Notification)
```

Where the first parameter is essentially the `withConnection` function that you had in your implementation. In substance this would be similar to, but much more general than 

```haskell
getNotification ::
  MVar PQ.Connection ->
  IO (Either IOError Notification)
```

Thus that library could then be used both inside "postgresql-simple" and to extend the functionality of such libraries as "hasql". This is separation of concerns and composition on library-scale in action. Of course, I imply that it would be best for you to do it, since it's your code after all.

nikita-volkov commented:
Oh, and I forgot to say that I'm sorry for taking so long to react to this PR, Leon.

lpsmith commented:
Oh, no problem at all on the time taken,  I am certainly in no rush.   (Though I don't speak for begriffs)

As for the "generic" getNotification,   that's not an interface that (A) anybody except for library implementors would want to use,  and would also imply that (B) HaSQL would have to expose some if its internals via an `Internal` module.    

Now,  while I fully support (B),   in which case you could have a separate hasql-notifications package if you really want,  I doubt that your proposed `getNotification` is really all that much more general,  as it's making an awful lot of assumptions as to what its `withConnection` parameter does and does not do;  for example,  it wouldn't really work for `withConnection` to take a connection out of a pool and put it back every time.

This construct is also isn't great from a code generation perspective,  as I doubt the object code for the "generalized" getNotification is really of much use in and of itself,   so the end-user executable doesn't really want it, and that you'd always want to inline that code (and also inline it's first parameter) in order to cut down on indirect jumps and whatnot.    Basically, it should be treated more like a macro than a higher-order function.   And getting that sort of transformation to work robustly with GHC is not worth the effort in this case.

Also,  I don't like this idea from the point of dependency hygiene on HaSQL's part or API hygiene on postgresql-libpq's part.

nikita-volkov commented:
> (B) HaSQL would have to expose some if its internals via an Internal module.

That's exactly what I propose to do on the Hasql side. E.g., I'd export a function like the following from the `Hasql.Connection` module:

```haskell
withLibPQConnection :: Hasql.Connection -> (PQ.Connection -> IO a) -> IO a
```

Then that would not only provide a way to solve this particular issue, but also a general way to provide other "libpq"-based extensions, without any pull-requests or any other kind of my involvement, which I find an incredibly important point from the maintenance perspective.

> that's not an interface that (A) anybody except for library implementors would want to use

Probably, but then we're talking about this in the context of a feature of which the same could be said just as well. I mean I've never used notifications in my life and I suspect the same could be said of most of the users. Actually, @begriffs is the only user requesting such hardcore features so far, and his case could be as well put in the "library implementors" group.

However, look at this:

```haskell
getNotification (withLibPQConnection hasqlConnection)
```

Apart from error-handling, that would be all it'd take to bind the proposed Hasql API with `getNotification` of the following signature:

```haskell
getNotification ::
  (forall a. (PQ.Connection -> IO a) -> IO a) ->
  IO (Either IOError Notification)
```

> I doubt that your proposed getNotification is really all that much more general, as it's making an awful lot of assumptions as to what its withConnection parameter does and does not do

It is more general, it does however indeed come with assumptions, so I think you meant to say that it's not safe. It's not safe indeed, but then it's almost the same level of abstraction as "libpq" itself that we're talking about. It does however solve a problem, which is common to separate families of libraries, and it doesn't impose any assumptions on how those libraries should wrap "libpq".

> This construct is also isn't great from a code generation perspective, as I doubt the object code for the "generalized" getNotification is really of much use in and of itself, so the end-user executable doesn't really want it, and that you'd always want to inline that code (and often also inline its first parameter) in the user-exported getNotification in order to cut down on indirect jumps and whatnot. Basically, it should be treated more like a macro than a higher-order function. And getting that sort of transformation to work robustly with GHC is not worth the effort in this case.

I don't get this concern. It's just a basic continuation. Even if it somehow will in fact impose any noticeable overhead it'll be trivially solvable with either the `INLINE` pragma or a combination of `INLINABLE` pragma and the `inline` function at the use site.

> Also, I don't like this idea from the point of dependency hygiene on hasql's part or API hygiene on postgresql-libpq's part.

I don't propose to change anything in "postgresql-libpq". I propose to release an extension library for it.

Concerning the "hasql" side, following is what concerns me most of all.

You see, lately I came to conclusion that the benefits of separation of concerns are not only applicable to modules, but the whole packages as well. And here's a specific practical motivation of mine: I'd prefer not to take on this functionality for maintenance, since I don't use it and don't plan to, hence there's not much sense for me to be responsible for it. I would much rather open the necessary doors in the API for it to be implementable as an extension by someone else, with him becoming both the copyright owner and the maintainer of that particular extension.

At the same time I'll keep my self able and the API flexible enough to continue the research in the area of the usability of the features, which this library already has. Those features are intentionally few: the library is basically just a mapping API, which can also execute statements. That's it. I'd love to be able to have the rest done in separate libraries and I provide more motivation on this [in the readme](https://github.com/nikita-volkov/hasql#benefits-of-being-an-ecosystem).

begriffs commented:
So should we go ahead with `withLibPQConnection` and create a hasql-notifications package? @lpsmith would that function give you what you need to implement your notification functions?

nikita-volkov commented:
I plan to expose that function either way.

lpsmith commented:
Yes, that function should be sufficient for my purposes.

lpsmith commented:
However,  while I'm 100% ok with such a solution,  the "generalized" `getNotification` that Nikita proposed still isn't very general:  it's basically assuming that it gives you exclusive access to a connection for a short duration.  So this should be considered part of the semantics of `withLibPQConnection`

mrijkeboer commented:
Any progress on this? I'm interested to use polling with HaSQL.

cocreature commented:
@lpsmith @begriffs Alright, we now have `withLibPQConnection` so there should be nothing in the way of a `hasql-notification` package. I’d be happy to help out or even convert your code into a package myself (giving appropriate credit ofc) but I don’t want to just “steal” it without your permission.

lpsmith commented:
The code is BSD3 licensed in postgresql-simple,  so you already have permission.   If you want explicit permission otherwise,   by all means,  please.  :)

cocreature commented:
FYI I have now created a [hasql-notifications](https://github.com/cocreature/hasql-notifications/) package. It’s not yet on hackage because there seems to be a race condition with `threadWaitRead` that does not occur using `postgresql-simple` but I hope I’ll be able to figure this out during the next days. (There is a testcase illustrating the race condition in case anybody else wants to look at it).

lpsmith commented:
I would be extremely interested in looking at the test case.   Cheers!

lpsmith commented:
Well, I'm not surprised that there is a race condition;  you removed the code to prevent several race conditions!

lpsmith commented:
Ok,  so the `join $ withLock fooLock $ do {...}` and the `join $ atomically $ do {...}` idioms are quite useful when dealing with concurrent programming;  the idiom allows you to decide what to do next while a lock is held (or in an atomic transaction!) and then the action that is returned is executed outside the lock/transaction.   (because,  for example,  the code returned may block,  or the code executes some IO actions that can't be performed safely inside STM.)

It's a highly useful idiom that I "discovered" myself,  but am by no means the first to discover it,  and this idiom is even used on hackage packages before I figured it out.  For example,  SPJ has at least once discussed this idiom on [haskell-cafe in 2006](https://mail.haskell.org/pipermail/haskell-cafe/2006-November/019771.html) and this idiom is also used in [resource-pool](https://hackage.haskell.org/package/resource-pool).

So by modifying code that perhaps looked like it was written in a needlessly obtuse way,  you are actually modifying the concurrent semantics of the code.  I discussed a very similar type of [race condition on my blog](https://blog.melding-monads.com/2015/11/21/announcing-linux-inotify-0-3/),  and also added comments explaining the cases to [postgresql-simple](https://hackage.haskell.org/package/postgresql-simple-0.5.2.1/docs/src/Database-PostgreSQL-Simple-Notification.html)

The only substantial difference between my [pull request](https://github.com/lpsmith/hasql/blob/c255be71acb00762e196112ce5f15dacd7223df7/library/Hasql/Notification.hs) is that the code returns `IOError`s instead of throws them.   I actually rewrote the notification handling code for hasql,  then backported the results back to postgresql-simple.

So,  what I really would do,  is stick to my PR and add the comments I added,  or stick with the postgresql-simple code and change it so that it returns IOErrors.   This will fix your errors.

I'm still extremely interested in a test case that demonstrates one of the race conditions!

cocreature commented:
@lpsmith Thanks a lot for looking at it so quickly. I’m not a big fan of the `join $ withLock` idiom since it makes it hard to see what is done inside and outside of the lock, but it is definitely better in this case.

However that is not what is causing the race condition, in fact I rewrote the `join` stuff in the hope of fixing a race condition that I might not have seen before. I am now back to the old code and enabled the code causing the race condition. The problem is in the third test. If I send the notification using `postgresql-simple`’s `exec` function (a comment illustrating why it is used instead of LibPQ.exec might be nice :)) everything works fine. However if I use a simple combination of `PQ.send` and `PQ.getResult` or hasql to send that notification, I sometimes run into a deadlock where no changes to the file descriptor are noticed.

begriffs commented:
Hey @nikita-volkov any interest in this PR?

Notifications would be the missing link that allows postgrest to reload its cache of the database structure automatically on changes (since we can create a DDL trigger that sends a NOTIFY to the server). The inability to auto refresh the cache has been a source of confusion for a long time.

If I understand correctly, the standalone hasql-notifications project ran into race condition problems that they could avoid if they were able to add something to this library to hook into the internals. @lpsmith told me he'd be willing to help move the feature forward in hasql if you're OK with it.

lpsmith commented:
Well, to be clear, @cocreature pointed out a race condition that I didn't fully appreciate, namely issues where `getNotification` blocks, then a notification arrives without GHC's IO manager becoming aware of it.  This frequently happens when a query causes a notification and the notification arrives in the same network packet with the result.

An acceptable fix is to add a `threadWaitRead` to a standard query.  This ensures that the IO manager stays in the loop, and thus wake up `getNotification` on every query.

`postgresql-simple` was vulnerable to this issue on Linux before I moved to async libpq calls for other reasons and thus inadvertently fixed this issue as well.

So it turns out that, as I have been arguing all along, there isn't really a generic way to implement `getNotification` for all consumers of `postgresql-libpq`.  It turns out that `getNotification` requires minor changes to the internal implementation details of hasql.

Also, it would probably be best to move `withLibpqConnection` into a module explicitly marked as Internal.  A libpq connection object is a complicated, stateful, very un-Haskelly thing, and making use of it without consideration of the internal implementation details of everything else using it is asking for trouble.

nikita-volkov commented:
The good news is that the coming native reimplementation based on the protocol will sport the Notification feature.

If anyone decides to bring this PR up to date and test it, I'll be willing to merge. However, beware that it's gonna go away in the next major release, because it'll be a complete reimplementation of the library with no Libpq dependency.

Also please notice that I am passionately against the `Internal` convention for plentiful reasons, so please don't move any modules there.

begriffs commented:
Very much looking forward to the native implementation! 🎊 I know it'll be "done when it's done," but any rough estimate of when you might release it?

nikita-volkov commented:
I plan to present the new version at [Haskell Exchange](https://skillsmatter.com/conferences/8522-haskell-exchange-2017#program) this year. So, some time before October 12.

nikita-volkov commented:
In case you're interested in the updates on the subject, [my talk from HaskellX](https://skillsmatter.com/skillscasts/10640-new-hasql-faster-and-simpler) is already published.

In short, the work is ongoing. The testing "beta" releases will from now on be in the `0.x` version space and the stable `0.19` has become `1`. The latest beta release is `0.20`, it natively implements the Postgres protocol, and it already supports Notification. However there still are issues (GC, in particular), which I'm working on, so it's not a major release yet.

begriffs commented:
@nikita-volkov I see that 0.20.1 is now available. Does it fix the GC issues you were talking about?

nikita-volkov commented:
@begriffs It's still in the works. Once it's production-ready it'll be released as v2.
